// react-docgen/dist/utils/parseJsDoc.js
/*
 * Copyright (c) 2015, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * 
 */

'use strict';

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});
exports['default'] = parseJsDoc;

var _doctrine = require('doctrine');

var _doctrine2 = _interopRequireDefault(_doctrine);

function getType(tag) {
  if (!tag.type) {
    return null;
  }
  return { name: tag.type.name };
}

// Add jsdoc @return description.
function getReturnsJsDoc(jsDoc) {
  var returnTag = jsDoc.tags.find(function (tag) {
    return tag.title === 'return' || tag.title === 'returns';
  });
  if (returnTag) {
    return {
      description: returnTag.description,
      type: getType(returnTag)
    };
  }
  return null;
}

// Add jsdoc @param descriptions.
function getParamsJsDoc(jsDoc) {
  if (!jsDoc.tags) {
    return [];
  }
  return jsDoc.tags.filter(function (tag) {
    return tag.title === 'param';
  }).map(function (tag) {
    return {
      name: tag.name,
      description: tag.description,
      type: getType(tag)
    };
  });
}

function parseJsDoc(docblock) {
  var jsDoc = _doctrine2['default'].parse(docblock);

  return {
    description: jsDoc.description || null,
    params: getParamsJsDoc(jsDoc),
    returns: getReturnsJsDoc(jsDoc)
  };
}

module.exports = exports['default'];
